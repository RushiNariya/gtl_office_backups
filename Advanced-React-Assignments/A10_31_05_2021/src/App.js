import React from 'react';
import logo from './logo_512.png';
import './App.css';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>Progressive Web Application</p>
      </header>
    </div>
  );
}

export default App;
