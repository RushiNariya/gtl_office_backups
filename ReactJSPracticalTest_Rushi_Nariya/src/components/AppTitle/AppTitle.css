.weather__logo {
  width: 40px;
  height: 40px;
}
.app__title {
  margin-top: 10px;
  margin-left: 10px;
}
