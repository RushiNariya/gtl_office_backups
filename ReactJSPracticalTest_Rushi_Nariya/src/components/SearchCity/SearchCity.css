@import 'https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css';

form {
  display: inline-block;
  margin: 1rem auto 0;
}

.pseudo-search {
  display: inline;
  position: relative;
  border: 2px solid #fff;
  border-radius: 5px;
  padding: 10px 15px;
  transition: 0.5 ease-in-out;
  outline: none;
}

input {
  border: 0;
  color: #fff;
  background-color: transparent;
  width: 400px;
}

input::placeholder {
  color: #fff;
  opacity: 1;
}

@media (max-width: 480px) {
  input {
    width: 240px;
  }
}

input:focus {
  outline: none;
}

button {
  color: white;
  border: none;
  background: none;
  cursor: pointer;
}

select {
  border: none;
}

.no-suggestions {
  color: #fff;
  padding: 0.5rem;
}

.suggestions {
  position: relative;
  border: 1px solid #999;
  color: #fff;
  font-size: 1.2rem;
  border-top-width: 0;
  list-style: none;
  margin-top: 5px;
  max-height: 143px;
  overflow-x: hidden;
  overflow-y: auto;
  padding-left: 0;
  width: inherit;
}

.suggestions::-webkit-scrollbar {
  display: none;
}
.suggestions li {
  padding: 0.5rem;
}

.suggestion-active,
.suggestions li:hover {
  background: rgba(255, 255, 255, 0.1);
  color: #fff;
  cursor: pointer;
  font-weight: 700;
}

.suggestions li:not(:last-of-type) {
  border-bottom: 1px solid #999;
}
