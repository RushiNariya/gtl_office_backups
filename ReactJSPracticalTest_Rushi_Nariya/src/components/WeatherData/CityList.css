.city-list {
  list-style-type: none;
  margin: 20px 20px 0;
  padding: 0;
  color: #fff;
}

.city-list .city-element {
  margin: 0;
  padding: 0;
}
