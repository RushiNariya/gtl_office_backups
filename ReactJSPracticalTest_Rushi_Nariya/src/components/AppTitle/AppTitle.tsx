import React from 'react';
import { Navbar } from 'react-bootstrap';
import weatherLogo from '../../assets/weather-app.png';
import './AppTitle.css';

function AppTitle(): JSX.Element {
  return (
    <Navbar bg-transparent="true" expand="lg">
      <Navbar.Brand href="#" className="text-white app__title">
        <span>
          <img src={weatherLogo} className="weather__logo" />
        </span>
        Weather App
      </Navbar.Brand>
    </Navbar>
  );
}

export default AppTitle;
