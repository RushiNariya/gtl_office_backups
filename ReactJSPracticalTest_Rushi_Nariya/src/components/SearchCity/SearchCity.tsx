import React, { useState, useContext, useEffect } from 'react';
import toast from 'react-hot-toast';
import suggestions from '../../Data/cities';
import { getSearchCityWeatherDataApi } from '../../Api/weatherAPI';
import './SearchCity.css';
import cityInfo from '../../intefaces/cityInfo';
import { GlobalContext } from '../../context/InitialState';

function SearchCity(): JSX.Element {
  const { searchCityHandler, errorHandler, resetSetErrorHandler } =
    useContext(GlobalContext);

  const initData: Array<{ id: number; city: string; }> = [];

  const [searchCity, setSearchCity] = useState('');
  const [showSuggestions, setShowSuggestions] = useState(false);
  const [filteredSuggestions, setFilteredSuggestions] = useState(initData);
  const [activeSuggestion, setActiveSuggestion] = useState(0);
  const [value, setValue] = useState('');

  const handleSubmit = (e): void => {
    e.preventDefault();
    setValue(searchCity);
    setSearchCity('');
  };

  const getWeatherByCity = async (): Promise<any> => {
    const data = await getSearchCityWeatherDataApi(value);

    if (data.cod === 200) {
      resetSetErrorHandler();

      const cityData: cityInfo = {
        name: data.name,
        country: data.sys.country,
        temp: data.main.temp,
        icon: data.weather[0].icon,
        longitude: data.coord.lon,
        latitude: data.coord.lat,
        maxTemp: data.main.temp_max,
        minTemp: data.main.temp_min,
      };
      searchCityHandler(cityData);
    } else {
      toast.error(data.message);
      errorHandler(data.message);
    }
    setSearchCity('');
  };

  useEffect(() => {
    if (value) {
      getWeatherByCity();
    }
  }, [value]);

  const onClick = (e): void => {
    setFilteredSuggestions([]);
    setShowSuggestions(false);
    setActiveSuggestion(0);
    setValue(e.currentTarget.innerText);
    setSearchCity('');
  };

  const onChange = (e): void => {
    setSearchCity(e.target.value);

    if (searchCity.length < 2) {
      setFilteredSuggestions([]);
      setShowSuggestions(false);
      setActiveSuggestion(0);
    }

    if (searchCity.length > 1) {
      const filteredSuggestions = suggestions.filter((suggestion) => {
        if (suggestion.city.toLowerCase().indexOf(searchCity.toLowerCase()) >= 0) {
          return suggestion;
        }
      });
      setFilteredSuggestions(filteredSuggestions);
      setShowSuggestions(true);
      setActiveSuggestion(0);
    }
  };

  const onKeyDown = (e): void => {
    // pressed the enter key
    if (e.keyCode === 13) {
      if (filteredSuggestions.length > 0) {
        setSearchCity(filteredSuggestions[activeSuggestion].city);
      } else {
        setSearchCity(e.target.value);
      }
      setShowSuggestions(false);
      setActiveSuggestion(0);
    }
    // pressed the up arrow
    else if (e.keyCode === 38) {
      if (activeSuggestion === 0) {
        return;
      }
      setActiveSuggestion((preValue) => preValue - 1);
    }
    // pressed the down arrow
    else if (e.keyCode === 40) {
      if (activeSuggestion - 1 === filteredSuggestions.length) {
        return;
      }
      setActiveSuggestion((preValue) => preValue + 1);
    }
  };

  let suggestionsListComponent;

  if (showSuggestions && searchCity) {
    if (filteredSuggestions.length) {
      suggestionsListComponent = (
        <ul className="suggestions">
          {filteredSuggestions.map((suggestion, index) => {
            let className;

            if (index === activeSuggestion) {
              className = 'suggestion-active';
            }
            return (
              <li className={className} key={suggestion.id} onClick={onClick}>
                {suggestion.city}
              </li>
            );
          })}
        </ul>
      );
    } else {
      suggestionsListComponent = (
        <div className="no-suggestions">
          <em>No suggestions, you are on your own!</em>
        </div>
      );
    }
  }

  return (
    <>
      <form onSubmit={handleSubmit}>
        <div className="pseudo-search">
          <input
            type="text"
            required
            placeholder="Search City"
            value={searchCity}
            onChange={onChange}
            onKeyDown={onKeyDown}
          />
          <button className="fa fa-search" type="submit"></button>
        </div>
        {suggestionsListComponent}
      </form>
    </>
  );
}
export default SearchCity;
