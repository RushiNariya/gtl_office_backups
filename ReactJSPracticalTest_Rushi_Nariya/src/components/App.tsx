import React from 'react';
import { Toaster } from 'react-hot-toast';
import AppTitle from './AppTitle/AppTitle';
import './App.css';
import WeatherData from './WeatherData/CityList';
import SearchCity from './SearchCity/SearchCity';

function App(): JSX.Element {
  return (
    <div className="App">
      <Toaster
        position="top-center"
        toastOptions={{
          duration: 5000,
          style: {
            borderRadius: '10px',
            background: '#fff',
            padding: '1rem',
            color: '#000',
          },
        }}
      />
      <AppTitle />
      <SearchCity />
      <WeatherData />
    </div>
  );
}

export default App;
