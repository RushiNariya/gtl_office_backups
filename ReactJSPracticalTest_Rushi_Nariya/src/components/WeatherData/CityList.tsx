import React, { useEffect, useContext } from 'react';
import toast from 'react-hot-toast';
import { getCurrentWeatherDataApi, getPosition } from '../../Api/weatherAPI';
import City from './CityData/City';
import './CityList.css';
import cityInfo from '../../intefaces/cityInfo';
import { GlobalContext } from '../../context/InitialState';

function WeatherData(): JSX.Element {
  const { searchCityHandler, errorHandler, resetSetErrorHandler, city } =
    useContext(GlobalContext);

  const getWeather = async (latitude: any, longitude: any): Promise<any> => {
    const data = await getCurrentWeatherDataApi(latitude, longitude);
    if (data.cod === 200) {
      resetSetErrorHandler();

      const cityData: cityInfo = {
        name: data.name,
        country: data.sys.country,
        temp: data.main.temp,
        icon: data.weather[0].icon,
        longitude: data.coord.lon,
        latitude: data.coord.lat,
        maxTemp: data.main.temp_max,
        minTemp: data.main.temp_min,
      };
      searchCityHandler(cityData);
    } else {
      errorHandler(data.message);
    }
  };

  useEffect(() => {
    getPosition()
      .then((position: any) => {
        getWeather(position.coords.latitude, position.coords.longitude);
        setInterval(
          () => getWeather(position.coords.latitude, position.coords.longitude),
          3600000
        );
      })
      .catch((err) => {
        toast.error(err.message);
      });
  }, []);

  return (
    <>
      {city.length ? (
        <ul className="city-list">
          {city.map((item) => {
            return (
                <li key={item.name} className="city-element">
                  <City cityData={item} />
                </li>
            );
          })}
        </ul>
      ) : <div className="city-list">No cities Found!</div>}
    </>
  );
}

export default WeatherData;
