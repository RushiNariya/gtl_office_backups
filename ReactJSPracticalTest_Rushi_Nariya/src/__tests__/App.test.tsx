import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import App from '../components/App';
import AppTitle from '../components/AppTitle/AppTitle';
import SearchCity from '../components/SearchCity/SearchCity';
import WeatherData from '../components/WeatherData/CityList';


let wrapped: ShallowWrapper;

beforeEach((): void => {
  wrapped = shallow(<App />);
});

it('renders the APP component', (): void => {
  expect(wrapped).toMatchSnapshot();
});

it('shows AppTitle', (): void => {
  expect(wrapped.find(AppTitle).length).toEqual(1);
});

it('shows SearchCity', (): void => {
  expect(wrapped.find(SearchCity).length).toEqual(1);
});

it('shows WeatherData', (): void => {
  expect(wrapped.find(WeatherData).length).toEqual(1);
});