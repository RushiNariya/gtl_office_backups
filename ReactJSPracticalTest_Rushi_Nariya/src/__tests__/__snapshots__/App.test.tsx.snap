// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the APP component 1`] = `
<div
  className="App"
>
  <Toaster
    position="top-center"
    toastOptions={
      Object {
        "duration": 5000,
        "style": Object {
          "background": "#fff",
          "borderRadius": "10px",
          "color": "#000",
          "padding": "1rem",
        },
      }
    }
  />
  <AppTitle />
  <SearchCity />
  <WeatherData />
</div>
`;

exports[`renders the component 1`] = `
<div
  className="App"
>
  <Toaster
    position="top-center"
    toastOptions={
      Object {
        "duration": 5000,
        "style": Object {
          "background": "#fff",
          "borderRadius": "10px",
          "color": "#000",
          "padding": "1rem",
        },
      }
    }
  />
  <AppTitle />
  <SearchCity />
  <WeatherData />
</div>
`;
